File: infrastructures/rest-ddd/csharp/src/Infrastructure/Repositories/Impl/ExampleCommandRepository.cs
/**
 * Example Command Repository Implementation
 * Handles write operations (CREATE, UPDATE, DELETE).
 */

using System;
using System.Threading.Tasks;
using Application.Mappers;
using Domain.Models;
using Infrastructure.Persistence.Write;

namespace Infrastructure.Repositories.Impl
{
    public class ExampleCommandRepository : ICommandRepository<ExampleModel, string>
    {
        /**
         * Command Repository for Example entity.
         * 
         * Rules:
         * - Located in /Infrastructure/Repositories/Impl/
         * - Suffix with 'Repository'
         * - Receives BMO, maps to WriteEntity internally
         * - Returns void or ID
         */

        private readonly ExampleMapper _mapper;
        // Add database connection/context here
        // private readonly DatabaseContext _dbContext;

        public ExampleCommandRepository(ExampleMapper p_mapper)
        {
            _mapper = p_mapper;
        }

        public async Task<string> SaveAsync(ExampleModel p_model)
        {
            // Map BMO → WriteEntity
            var entity = _mapper.ToWriteEntity(p_model);

            // Save to database
            // await _dbContext.Examples.InsertOneAsync(entity);

            // Placeholder: generate ID
            entity.Id = GenerateId();

            return entity.Id;
        }

        public async Task UpdateAsync(ExampleModel p_model)
        {
            // Map BMO → WriteEntity
            var entity = _mapper.ToWriteEntity(p_model);

            // Update version for optimistic locking
            entity.IncrementVersion();

            // Update in database
            // var filter = Builders<ExampleWriteEntity>.Filter.And(
            //     Builders<ExampleWriteEntity>.Filter.Eq(e => e.Id, entity.Id),
            //     Builders<ExampleWriteEntity>.Filter.Eq(e => e.Version, entity.Version - 1)
            // );
            // await _dbContext.Examples.ReplaceOneAsync(filter, entity);

            await Task.CompletedTask;
        }

        public async Task DeleteAsync(string p_id)
        {
            // Delete from database
            // await _dbContext.Examples.DeleteOneAsync(e => e.Id == p_id);

            await Task.CompletedTask;
        }

        public async Task<bool> ExistsAsync(string p_id)
        {
            // Check if exists in database
            // var count = await _dbContext.Examples.CountDocumentsAsync(e => e.Id == p_id);
            // return count > 0;

            // Placeholder
            await Task.CompletedTask;
            return false;
        }

        private string GenerateId()
        {
            return $"example_{DateTime.UtcNow.Ticks}_{Guid.NewGuid().ToString("N").Substring(0, 8)}";
        }
    }
}

File: infrastructures/rest-ddd/typescript/src/infrastructure/repositories/impl/ExampleCommandRepository.ts
/**
 * Example Command Repository Implementation
 * Handles write operations (CREATE, UPDATE, DELETE).
 */

import { ICommandRepository } from '../ICommandRepository';
import { ExampleModel } from '../../../domain/models/ExampleModel';
import { ExampleWriteEntity } from '../../persistence/write/ExampleWriteEntity';
import { ExampleMapper } from '../../../application/mappers/ExampleMapper';

export class ExampleCommandRepository implements ICommandRepository<ExampleModel, string> {
  /**
   * Command Repository for Example entity.
   * 
   * Rules:
   * - Located in /infrastructure/repositories/impl/
   * - Suffix with 'Repository'
   * - Receives BMO, maps to WriteEntity internally
   * - Returns void or ID
   */

  constructor(
    private readonly _mapper: ExampleMapper
    // Add database connection/context here
    // private readonly _dbContext: DatabaseContext
  ) {}

  async saveAsync(p_model: ExampleModel): Promise<string> {
    // Map BMO → WriteEntity
    const entity = this._mapper.toWriteEntity(p_model);

    // Save to database
    // const savedEntity = await this._dbContext.examples.insertOne(entity);
    // entity.id = savedEntity.insertedId.toString();

    // Placeholder: generate ID
    entity.id = this.generateId();

    return entity.id;
  }

  async updateAsync(p_model: ExampleModel): Promise<void> {
    // Map BMO → WriteEntity
    const entity = this._mapper.toWriteEntity(p_model);

    // Update version for optimistic locking
    entity.incrementVersion();

    // Update in database
    // await this._dbContext.examples.updateOne(
    //   { _id: entity.id, version: entity.version - 1 },
    //   { $set: entity }
    // );
  }

  async deleteAsync(p_id: string): Promise<void> {
    // Delete from database
    // await this._dbContext.examples.deleteOne({ _id: p_id });
  }

  async existsAsync(p_id: string): Promise<boolean> {
    // Check if exists in database
    // const count = await this._dbContext.examples.countDocuments({ _id: p_id });
    // return count > 0;

    // Placeholder
    return false;
  }

  private generateId(): string {
    return `example_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }
}
